woocommerce/woocommerce
# Marketplace: Add "Add to Store" button for free and WordPress.org products

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:

- Allows merchants to install any free and .org hosted product from any page, while creating the order on the go. Currently this mocks order creation process.
- Adds new `<InstallNewProductModal>` component. This listens for changes in `query`, and shows the corresponding modal to the user.
- Adds new **Add to Store** button on the search pages. You need to switch to the Woo.com branch as well.




### How to test the changes in this Pull Request:

1. Connect front end and back end to your local Woo.com instance. 
    - For front end, switch `MARKETPLACE_HOST` in `marketplace/components/constants.ts` to `https://woocommerce.test`.
    - For back-end, make modifications in `class-wc-helper`. Make sure you also connect to the right Docker network.
2. Build assets with `pnpm --filter='@woocommerce/plugin-woocommerce' watch:build`
3. Disconnect store from Woo.com by clicking **Avatar > Disconnect** from store
    - Please note which page you initiate the disconnection from. After the disconnect is done, you should be redirected back to the same page.
5. [On the search page](http://localhost:8888/wp-admin/admin.php?page=wc-admin&tab=extensions&path=%2Fextensions), find a product with the **Add to Store** button and click it. It should show a warning for requiring a connection.
6. Complete the connection. Please note when you come back to your store, the modal should open automatically.
7. In the next modal, click Install. This will first ping Woo.com, create an order and then start downloading the product from WordPress.org repository. 
8. When the installation is done and we're allowed to activate (sometimes we're not allowed), you should see the **Activate** button. Clicking the button will activate the product and then show an option to see the plugin list, to set up the plugin. 
9. On Woo.com, delete your customer's billing address and try to install the plugin again. You won't see the **Add to Store** on the same product again, because it's now installed. 
10. Trying to install a new product, should first show a warning and then redirect user to Woo.com with the product added to cart.
11. When it's done, we should test installations with a product key. Find a product key from the API requests in [the My Subscriptions page](http://localhost:8888/wp-admin/admin.php?page=wc-admin&tab=my-subscriptions&path=%2Fextensions) and then visit http://localhost:8888/wp-admin/admin.php?page=wc-admin&tab=my-subscriptions&path=%2Fextensions&install=PRODUCT_KEY_STARTS_WITH_W00.
12. On the [the My Subscriptions page](http://localhost:8888/wp-admin/admin.php?page=wc-admin&tab=my-subscriptions&path=%2Fextensions) click **Install** in the table and confirm it works.
